cmake .. Returns error libatomic not found

Hi I cannot install libmultiprocess

I am returned the error:

CMake Error at /usr/lib/x86_64-linux-gnu/cmake/CapnProto/CapnProtoConfig.cmake:71 (message):
  libatomic not found
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)

I have installed capnproto from master tree c++ here:

https://github.com/capnproto/capnproto/tree/master/c%2B%2B

using autoreconf -i && ./configure && make -j6 check && sudo make install

Running CMake .. on build directory still did not work.

I double checked libatomic-ops and installed from master tree at

https://github.com/ivmai/debian-libatomic-ops

I updated libstdc++ with sudo apt-get install libstdc++ and apt -fix-broken install for safe measure

I again ran for capnproto on master/capnproto/c++

autoreconf -i && ./configure && make -j6 check && sudo make install

to see if the reconfiguration would link libatomic

Again testing CMake .. on libmultiprocess/build I am returned:

CMake Error at /usr/lib/x86_64-linux-gnu/cmake/CapnProto/CapnProtoConfig.cmake:71 (message):
  libatomic not found
Call Stack (most recent call first):
  CMakeLists.txt:9 (find_package)

Why is CMake not finding libatomic and connecting it for capnproto?

Under capnproto issues another user had a similar problem with Raspbian. In my case I am running Kali with Debian Bullseye repositories overtop. My system is well built with GNU libraries as well, and runs as a full fleshed out Debian distro inside a Kali. I am on linux x86_64.

https://github.com/capnproto/capnproto/issues/1448

In his case capnproto did not want to build at all, but it passes with all tests succeeded in my case, however it is not liking to configure with libatomic-ops, even though it is already installed.

Because of this the CMake for libmultiprocess cannot be done.

Just looking to set libmultiprocess in place for installing bitcoin and master tree using --enable-libmultiprocess for added compatibility in advance of future needs. I don't want to have to recomp[ile with enable multiprocess later on, it could be frustrating. I am aware that this is entirely to assist in the C++ libraries for compiling purposes only, and doesn't really have any major affect on bitcoin-core installation, but I would prefer using libmultiprocess with bitcoin for any compatibility issues post installation or with accessory uses.

I would appreciate the time looked into this.

Satori Hoshi

Wow, what a journey. I was able to reproduce the issue on debian by just trying to use build libmultiprocess against current libcapnp-dev package. Error is

```
-- Looking for __atomic_load_8 in atomic
-- Looking for __atomic_load_8 in atomic - not found
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/CapnProto/CapnProtoConfig.cmake:71 (message):
  libatomic not found
```

from 

```cmake
if (yes)  # WITH_LIBATOMIC
  include(CheckLibraryExists)
  check_library_exists(atomic __atomic_load_8 "" FOUND_LIBATOMIC)
  if (FOUND_LIBATOMIC)
    list(APPEND CMAKE_REQUIRED_LIBRARIES "atomic")
  else()
    message(FATAL_ERROR "libatomic not found")
  endif()
endif()
```

which seems to come from this debian patch https://sources.debian.org/patches/capnproto/0.9.1-4/07_libatomic.patch/ unnecessarily requiring libatomic.

The error seems to happen because `/usr/lib/x86_64-linux-gnu/libatomic.so.1.2.0` from https://packages.debian.org/sid/libatomic1 doesn't have an `__atomic_load_8` symbol, but I'm not sure how `WITH_LIBATOMIC` autoconf variable would have been yes in that case during the capnproto package build

Will look into this more and try to find a workaround. I'd prefer the build just work and not have to update the instructions to work around it.

Figured out the root cause. It is a bug in the debian package caused by a limitation of cmake. Cmake fails to find the `__atomic_load_8` symbol even though it exists. Running cmake with `cmake --debug-trycompile ..` shows:

```c++
Building CXX object CMakeFiles/cmTC_6fce7.dir/CheckFunctionExists.cxx.o
/usr/bin/c++   -DCHECK_FUNCTION_EXISTS=__atomic_load_8 -o CMakeFiles/cmTC_6fce7.dir/CheckFunctionExists.cxx.o -c /root/libmultiprocess/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx
<command-line>: error: new declaration ‘char __atomic_load_8()’ ambiguates built-in declaration ‘long unsigned int __atomic_load_8(const volatile void*, int)’ [-fpermissive]
<command-line>: note: in definition of macro ‘CHECK_FUNCTION_EXISTS’
/root/libmultiprocess/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx: In function ‘int main(int, char**)’:
/root/libmultiprocess/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx:17:24: error: too few arguments to function ‘long unsigned int __atomic_load_8(const volatile void*, int)’
   17 |   CHECK_FUNCTION_EXISTS();
gmake[1]: *** [CMakeFiles/cmTC_6fce7.dir/build.make:78: CMakeFiles/cmTC_6fce7.dir/CheckFunctionExists.cxx.o] Error 1
```

Looking at the `CheckFunctionExists.cxx` test program shows it can't compile because it tries to declare the `__atomic_load_8` function in the global C++ namespace, and this conflicts with the compiler built in defintion. Autoconf avoids this problem by declaring the function inside a `namespace conftest` c++ namespace, and fixing the symbol name by declaring it `extern "C"`

So autoconf is able to find the symbol and cmake is unable to find it. So the cmake code debian added in https://sources.debian.org/patches/capnproto/0.9.1-4/07_libatomic.patch/ is broken, and probably needs to be fixed to detect the the libatomic library another way, or cmake `check_library_exists` function needs to be improved and maybe use the same namespace trick that autoconf uses.

Either way I think hardcoding `set(FOUND_LIBATOMIC TRUE)` in the CMakeLists.txt file might be the easiest workaround, and maybe some of @SatoriHoshiAiko's documentation suggestions would help as well. Will follow up soon with some fix

Well the workaround for me was simple.

Not use the Debian package version, and just build from the capnproto
source.

I have a tendency to try every available option when I have an issue, to
ensure it is fully corrected. I can confirm what you are seeing is correct,
the Debian apt repository has this error.

Once I saw that I checked the same config.cmake of the capnproto source
version, it was not there.

Building from their source resolved the issue and libmultitool built with
no issue.

I think this is the most direct solution. Except for perhaps posting the
issue to where the Debian package is built and notifying that libatomic
should not be unilaterally requested.

But to me it was just uninstall one (Debian) and use the other (CapnProto
git source). Problem solved.

SatoriHoshiAiko

On Mon, May 23, 2022 at 1:58 PM Ryan Ofsky ***@***.***> wrote:

> Wow, what a journey. I was able to reproduce the issue on debian by just
> trying to use build libmultiprocess against current libcapnp-dev package.
> Error is
>
> -- Looking for __atomic_load_8 in atomic
> -- Looking for __atomic_load_8 in atomic - not found
> CMake Error at /usr/lib/x86_64-linux-gnu/cmake/CapnProto/CapnProtoConfig.cmake:71 (message):
>   libatomic not found
>
> from
>
> if (yes)  # WITH_LIBATOMIC
>   include(CheckLibraryExists)
>   check_library_exists(atomic __atomic_load_8 "" FOUND_LIBATOMIC)
>   if (FOUND_LIBATOMIC)
>     list(APPEND CMAKE_REQUIRED_LIBRARIES "atomic")
>   else()
>     message(FATAL_ERROR "libatomic not found")
>   endif()endif()
>
> which seems to come from this debian patch
> https://sources.debian.org/patches/capnproto/0.9.1-4/07_libatomic.patch/
> unnecessarily requiring libatomic.
>
> The error seems to happen because
> /usr/lib/x86_64-linux-gnu/libatomic.so.1.2.0 from
> https://packages.debian.org/sid/libatomic1 doesn't have an __atomic_load_8
> symbol, but I'm not sure how WITH_LIBATOMIC autoconf variable would have
> been yes in that case during the capnproto package build
>
> Will look into this more and try to find a workaround. I'd prefer the
> build just work and not have to update the instructions to work around it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/chaincodelabs/libmultiprocess/issues/68#issuecomment-1135083069>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWUI5A7MLKJH3JCGBNTIDVLVLPPPVANCNFSM5WTASUSQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

I started encountering this error after upgrading from Ubuntu 22.04 to 23.04. However, I didn't try right before the upgrade, so perhaps that's unrelated.

```
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "3.0.8") found components: Crypto SSL 
-- Looking for __atomic_load_8 in atomic
-- Looking for __atomic_load_8 in atomic - not found
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/CapnProto/CapnProtoConfig.cmake:71 (message):
  libatomic not found
Call Stack (most recent call first):
  CMakeLists.txt:23 (find_package)
```

Using capnproto 0.9.2-2.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,6 +8,8 @@ project("Libmultiprocess" CXX)
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED YES)
 
+include("cmake/compat_find.cmake")
+
 find_package(CapnProto REQUIRED)
 find_package(Threads REQUIRED)
 
@@ -20,7 +22,7 @@ if(Libmultiprocess_ENABLE_CLANG_TIDY)
   set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_EXECUTABLE}")
 endif()
 
-include("cmake/capnp_compat.cmake")
+include("cmake/compat_config.cmake")
 include("cmake/pthread_checks.cmake")
 include(GNUInstallDirs)
 

--- a/cmake/compat_config.cmake
+++ b/cmake/compat_config.cmake
@@ -2,6 +2,9 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+# compat_config.cmake -- compatibility workarounds meant to be included after
+# cmake find_package() calls are made, before configuring the ebuild
+
 # Define capnp_PREFIX if not defined to avoid issue on macos
 # https://github.com/chaincodelabs/libmultiprocess/issues/26
 

--- a/cmake/compat_find.cmake
+++ b/cmake/compat_find.cmake
@@ -1,0 +1,20 @@
+# Copyright (c) 2024 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+# compat_find.cmake -- compatibility workarounds meant to be included before
+# cmake find_package() calls are made
+
+# Set FOUND_LIBATOMIC to work around bug in debian capnproto package that is
+# debian-specific and does not happpen upstream. Debian includes a patch
+# https://sources.debian.org/patches/capnproto/1.0.1-4/07_libatomic.patch/ which
+# uses check_library_exists(atomic __atomic_load_8 ...) and it fails because the
+# symbol name conflicts with a compiler instrinsic as described
+# https://github.com/chaincodelabs/libmultiprocess/issues/68#issuecomment-1135150171.
+# This could be fixed by improving the check_library_exists function as
+# described in the github comment, or by changing the debian patch to check for
+# the symbol a different way, but simplest thing to do is work around the
+# problem by setting FOUND_LIBATOMIC. This problem has probably not
+# been noticed upstream because it only affects CMake packages depending on
+# capnproto, not autoconf packages.
+set(FOUND_LIBATOMIC TRUE)

